File: uponor/drivers/thermostat/driver.js
'use strict';

const Homey = require('homey');
const { getVars, parseThermostats } = require('../../lib/uponor');

module.exports = class ThermostatDriver extends Homey.Driver {
  async onInit() {
    this.log('Uponor thermostat driver initialized');
  }

  async onPair(session) {
    let host = '';

    session.setHandler('set_host', async (value) => {
      host = String(value || '').trim();
      if (!host) throw new Error('Enter the gateway IP address');
      await getVars(host); // fail fast, before moving to the device list
      return true;
    });

    session.setHandler('list_devices', async () => {
      const vars = await getVars(host);
      const rooms = parseThermostats(vars);
      if (rooms.length === 0) {
        throw new Error('No thermostats found on that gateway');
      }
      return rooms.map((room) => ({
        name: room.name || `Room C${room.controller}T${room.thermostat}`,
        data: { id: room.prefix },
        store: { host, controller: room.controller, thermostat: room.thermostat },
        capabilities:
          room.humidity !== null
            ? ['measure_temperature', 'measure_humidity']
            : ['measure_temperature'],
      }));
    });
  }
};

File: uponor-smatrix_x-265+r-208/drivers/thermostat/driver.py
from homey.driver import Driver, ListDeviceProperties
from homey.pair_session import PairSession

from ...lib.uponor import get_vars, parse_thermostats


class ThermostatDriver(Driver):
    async def on_init(self):
        await super().on_init()
        self.log("Uponor thermostat driver initialized")

    async def on_pair(self, session: PairSession):
        host = ""

        async def set_host(value):
            nonlocal host
            host = str(value or "").strip()
            if not host:
                raise ValueError("Enter the gateway IP address")
            await get_vars(host)  # fail fast, before moving to the device list
            return True

        async def list_devices(_view_data) -> list[ListDeviceProperties]:
            rooms = parse_thermostats(await get_vars(host))
            if not rooms:
                raise ValueError("No thermostats found on that gateway")
            return [
                {
                    "name": room["name"] or f"Room C{room['controller']}T{room['thermostat']}",
                    "data": {"id": room["prefix"]},
                    "store": {
                        "host": host,
                        "controller": room["controller"],
                        "thermostat": room["thermostat"],
                    },
                    "capabilities": (
                        ["measure_temperature", "measure_humidity"]
                        if room["humidity"] is not None
                        else ["measure_temperature"]
                    ),
                }
                for room in rooms
            ]

        session.set_handler("set_host", set_host)
        session.set_handler("list_devices", list_devices)


homey_export = ThermostatDriver
